### ISSUE TITLE
Production grade deployments

### ISSUE THREAD

**Issue by lgalabru** [hirosystems/clarinet#312]
Something that has been in the back of my mind for a while now, that would address a few deployment related feature requests, the right way.
I keep seeing ad-hoc deployment scripts in clarinet repos. 
Deployments scripts certainly work, but they are repetitive, error prone, and don't help inspiring confidence in my mind.
I'd love to see deployments being redesigned, based on DevOps best practices. 
As of today, developers can use 
```
clarinet contract publish
```
to publish their whole protocol, but this approach is limited, and is not always useful or relevant when you need to deploy:
- large protocols
- protocol upgrades

I had an intensive experience with [Terraform](https://www.terraform.io) in the past, and would like to migrate deployments to a similar 2 steps process:
```
clarinet deployment plan
```
would create a descriptive deployment plan (materialized as a YAML or HCL file, in `deployments/<date>.yaml`), based on previous deployments, if any.
By default, if it's your first deployment, you'd get something like:

```yaml

contract_publish:
    - deployer: "deployer"
    - name: "my-contract"
    - hash: "0x000....01"
    - version: "v1"

```

Developers would be able to edit and commit this files, which would be helpful if some `contract-call` need to be integrated in the deployment plan:

```yaml
contract_publish:
    - deployer: "deployer"
    - name: "my-contract"
    - version: "v1"
    - hash: "0x000....01"

contract_call:
    - contract_id: "deployer. my-contract-v1"
    - method: "say-hi"
    - arguments:
         - u1
```

Later on, if `my-contract` is edited (and v1 deployed), a new plan would detect the changes and would suggest:

```yaml
contract_publish:
    - deployer: "deployer"
    - name: "my-contract"
    - version: "v2"
    - hash: "0x000....02"
```

Once deployment plans are validated and ready, they can be tested and ran with:

```
clarinet deployment run 
```

I think that this approach would help us getting rid of deployment scripts, but I'd be curious to get some feedbacks from community members.

**Comment by lgalabru** [comment]
In terms of implementation, it should not be too bad, as the latest approach in the code is already performing some sort of 2 steps process.

**Comment by lgalabru** [comment]
Addressed in #333
